### DISCUSSION TITLE
Completing a Purchase with Stripe

### DISCUSSION THREAD

**Discussion by JS00001** [saleor/saleor#11845] [Q&A]
Hello,

I've been combing through the Saleor docs (following these: https://docs.saleor.io/docs/3.x/developer/available-plugins/stripe and https://docs.saleor.io/docs/3.x/developer/checkout). However, I have found what seems to be an issue (maybe just my ignorance). 

This is my *rough* frontend code, I'll offer a short walkthrough

```javascript
const onCompletePurchase = async () => {
    createPaymentIntentMutation.mutate(undefined, {
      onSuccess: () => {
        completePaymentMutation.mutate(undefined, {
          onSuccess: async ({ data }) => {
            console.log(data);
            const clientSecret = JSON.parse(data.confirmationData as string).client_secret;

            confirmPayment(clientSecret, {
              paymentMethodType: "Card",
            }).then((result) => {
              const { error, paymentIntent } = result;

              if (error) {
                console.log(error);
              } else if (paymentIntent) {
                console.log(JSON.stringify(paymentIntent, null, 2));
                completePaymentMutation.mutate(undefined, {
                  onSuccess: () => {
                    navigation.navigate("CartLanding");
                    setCheckout({} as Checkout);
                  },
                  onError: (error) => {
                    console.log(error);
                  },
                });
              } else {
                console.log("Something went wrong");
              }
            });
          },
          onError: (error) => {
            console.log(error);
          },
        });
      },
      onError: (error) => {
        console.log(error);
      },
    });
  };
```
**Walkthrough** - *(Follows these steps https://docs.saleor.io/docs/3.x/developer/available-plugins/stripe)*
- Creates the payment intent (through saleor)
- Completes the payment to get the confirmation information IE: Client Secret (through saleor)
- Once gets Client Secret, confirms the payment through stripe
- **Where issue occurs** - Tries to get the order from completing the payment again as detailed in the docs as below
```
Finalizing a payment
When the customer finalizes a payment, Stripe sends a webhook notification with all details related to this payment. Saleor asynchronously receives the notification and completes checkout process.

On client-side, when stripe.confirmCardPayment doesn't return an error as [described here](https://stripe.com/docs/payments/accept-a-payment?platform=web&ui=elements#web-submit-payment),
client-side can call checkoutComplete one more time. Saleor confirms that the payment has success status and returns an order object.
```

However, I believe the issue is this. I call "checkoutComplete" at the same time that saleor receives the webhook notification, resulting in an error with the clients request. If I do for example, setTimeout for 2 seconds, and THEN call checkoutComplete, it works fine, but I believe there is an issue with both of these being called. Am I missing something? or is  there a better way to do this WITHOUT using a setTimeout?

Thanks and feel free to ask some questions to figure out the specifics.

**Comment by timuric** [comment]
@JS00001, have you considered using stripe app instead of plugin? https://github.com/saleor/react-storefront

**Comment by tanujrawat95** [comment]
Hi @JS00001 , I am also trying to integrate stripe with saleor. We are trying to keep the saleor core graphql apis and the dashboard page. We will create our own strore-front UI. I have completed the integration till i get the client secret after creating payment intent. Now I am stuck how to make payments and get webhooks( also whhat will be the endpoint for that). Please help me out with that
